File: backend/mssaga/src/main/java/mssaga/mssaga/controllers/CadastroController.java
package mssaga.mssaga.controllers;

import mssaga.mssaga.DTO.ClienteDTO;
import mssaga.mssaga.sagas.AutocadastroSaga;

import java.util.HashMap;
import java.util.Map;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/clientes")
public class CadastroController {

    @Autowired
    private AutocadastroSaga autocadastroSaga;

    @PostMapping
    public ResponseEntity<Map<String, Object>> cadastrarCliente(@RequestBody ClienteDTO clienteDTO) {
        try {
            ClienteDTO clienteCriado = autocadastroSaga.autoCadastro(clienteDTO);
            Map<String, Object> response = new HashMap<>();
            response.put("cpf", clienteCriado.getCpf());
            response.put("email", clienteCriado.getEmail());
            response.put("codigo", clienteCriado.getCodigo());
            return ResponseEntity.status(201).body(response);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.status(409).body(Map.of("erro", e.getMessage()));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(Map.of("erro", "Erro ao processar autocadastro"));
        }
    }

}

File: backend/saga-orchestration-service/src/main/kotlin/br/ufpr/dac/saga_orchestration_service/controllers/CadastroController.kt
package br.ufpr.dac.saga_orchestration_service.controllers

import br.ufpr.dac.saga_orchestration_service.sagas.AutocadastroSaga
import br.ufpr.dac.saga_orchestration_service.sagas.CadastroFuncionarioSaga
import jakarta.validation.Valid
import kotlinx.coroutines.runBlocking
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.CrossOrigin
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import utils.dto.ClienteInputDTO
import utils.dto.ClienteOutputDTO
import utils.dto.FuncionarioInputDTO
import utils.dto.FuncionarioOutputDTO
import java.net.URI
import org.springframework.beans.factory.annotation.Value

@RestController
@RequestMapping("/v1")
@CrossOrigin(origins = ["\${gateway.url}"])
class CadastroController(
    private val autocadastroSaga: AutocadastroSaga,
    private val cadastroFuncionarioSaga: CadastroFuncionarioSaga
) {

    @PostMapping("/clientes")
    fun autocadastro(@RequestBody @Valid body: ClienteInputDTO): ResponseEntity<ClienteOutputDTO> {
        val cliente = runBlocking {
            autocadastroSaga.executeSaga(body)
        }

        return ResponseEntity.created(URI("/clientes/${cliente.codigo}")).body(cliente)
    }

    @PostMapping("/funcionarios")
    fun cadastrarFuncionario (@RequestBody @Valid body: FuncionarioInputDTO): ResponseEntity<FuncionarioOutputDTO> {
        val funcionario = runBlocking {
            cadastroFuncionarioSaga.executeSaga(body)
        }

        return ResponseEntity.created(URI("/funcionarios/${funcionario.codigo}")).body(funcionario)
    }

}